### PR TITLE
added mvs implementation

### PR DESCRIPTION
**Pull request by ibr11**
Before submitting a pull request, please do the following steps:

1. Read instructions for contributors [here](https://tech.yandex.com/catboost/doc/dg/concepts/development-and-contributions-docpage/).
2. Run `ya make` in catboost folder to make sure the code builds.
3. Add tests that test your change.
4. Run tests using `ya make -t -A` command.
5. If you haven't already, complete the CLA.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=ru.

### CODE REVIEW

**Comment by annaveronika** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@ibr11, please agree to the CLA (https://github.com/catboost/catboost/blob/master/CONTRIBUTING.md#provide-contributions)

**Comment by arcadia-devtools** [comment]
@annaveronika, internal review request created: [961303](https://a.yandex-team.ru/review/961303)

File: catboost/libs/algo/mvs.cpp [hunk 1]
```diff
@@ -18,22 +18,110 @@ inline static double GetSingleProbability(double derivativeAbsoluteValue, double
...
+                    gradSumInBlock[blockId] += Abs(*it);
+                }
+            },
+            0,
+            blockParams.GetBlockCount(),
+            NPar::TLocalExecutor::WAIT_COMPLETE
+        );
+
+        const double sum_gradients = Accumulate(gradSumInBlock.begin(), gradSumInBlock.end(), 0.0);
+        return sum_gradients / cnt;
```
Comment (annaveronika) [inline catboost/libs/algo/mvs.cpp]: Please fix the style guide - C++ code uses camelCase naming for local variables.

File: catboost/libs/algo/mvs.cpp [hunk 2]
```diff
@@ -18,22 +18,110 @@ inline static double GetSingleProbability(double derivativeAbsoluteValue, double
...
+}
+
+double TMvsSampler::GetLambda(
+    const double* derivatives,
```
Comment (annaveronika) [inline catboost/libs/algo/mvs.cpp]: style guilde again - output variables should be last ones among other parameters.
Also if something is changed inside the function (in this case derivatives), it should be obvious from the interface. In this case it's not clear.

File: catboost/libs/algo/mvs.cpp [hunk 3]
```diff
@@ -18,22 +18,110 @@ inline static double GetSingleProbability(double derivativeAbsoluteValue, double
...
+        ? CalculateLastIterMeanLeafValue(leafValues)
+        : CalculateMeanGradValue(derivatives, SampleCount, localExecutor);
+    return mean * mean;
+}
+
+double TMvsSampler::CalculateThreshold(
+    TVector<double>::iterator candidatesBegin,
+    TVector<double>::iterator candidatesEnd,
+    double sumSmall,
+    ui32 nLarge,
```
Comment (annaveronika) [inline catboost/libs/algo/mvs.cpp]: We are not using single letter variable names in code (in this case n), please give it a more meaningful name.

File: catboost/libs/algo/mvs.h [hunk 4]
```diff
@@ -15,19 +16,34 @@ namespace NPar {
 
 class TMvsSampler {
 public:
-    TMvsSampler(ui32 sampleCount, float headFraction)
+    TMvsSampler(ui32 sampleCount, float sampleRate, float lambda, bool lambdaIsSet)
```
Comment (annaveronika) [inline catboost/libs/algo/mvs.h]: You can use TMaybe instead of lambdaIsSet

File: catboost/libs/algo/tensor_search_helpers.cpp [hunk 5]
```diff
@@ -363,7 +364,8 @@ void Bootstrap(
     const float baggingTemperature = params.ObliviousTreeOptions->BootstrapConfig->GetBaggingTemperature();
     const float takenFraction = params.ObliviousTreeOptions->BootstrapConfig->GetTakenFraction();
     const bool isPairwiseScoring = IsPairwiseScoring(params.LossFunctionDescription->GetLossFunction());
-    const float headFraction = params.ObliviousTreeOptions->BootstrapConfig->GetMvsHeadFraction();
+    const float mvsReg = params.ObliviousTreeOptions->BootstrapConfig->GetMvsReg();
+    const bool mvsRegIsSet = params.ObliviousTreeOptions->BootstrapConfig->MvsRegIsSet();
     bool isCoinFlipping = true;
```
Comment (annaveronika) [inline catboost/libs/algo/tensor_search_helpers.cpp]: please rename isCoinFlipping to useRandomSampling or something else, which would be meaningful

File: catboost/libs/algo/mvs.cpp (hunk 2 above)
Comment (ibr11) [reply catboost/libs/algo/mvs.cpp]: I guess, you did not notice "const". Or is the pointer to constant an output variable?

**Comment by arcadia-devtools** [comment]
Check [961303](https://a.yandex-team.ru/review/961303) failed: 6 build(s) broken, 14 test(s) broken

File: catboost/libs/algo/mvs.cpp (hunk 2 above)
Comment (annaveronika) [reply catboost/libs/algo/mvs.cpp]: You're right, I didn't notice const, closing this one.

**Comment by annaveronika** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@annaveronika, internal review request updated: [961303](https://a.yandex-team.ru/review/961303)

**Comment by arcadia-devtools** [comment]
Check [961303](https://a.yandex-team.ru/review/961303) failed: 6 build(s) broken, 12 test(s) broken

**Comment by arcadia-devtools** [comment]
CLA already signed

**Comment by annaveronika** [comment]
@arcadia-devtools Ship it!

File: catboost/pytest/test.py [hunk 6]
```diff
@@ -7011,7 +7011,7 @@ def run_catboost(eval_path, mvs_head_fraction):
             '--has-header',
             '--boosting-type', boosting_type,
             '--bootstrap-type', 'MVS',
-            '--mvs-head-fraction', mvs_head_fraction,
+            '--subsample', mvs_sample_rate,
```
Comment (annaveronika) [inline catboost/pytest/test.py]: Could you add mvs parameters to python-package/catboost/core.py to CatBoostClassifier and CatBoostRegressor __init__ function?

**Comment by annaveronika** [comment]
@arcadia-devtools Ship it!

**Comment by annaveronika** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@annaveronika, internal review request updated: [961303](https://a.yandex-team.ru/review/961303)

**Comment by arcadia-devtools** [comment]
Check [961303](https://a.yandex-team.ru/review/961303) failed: 6 build(s) broken, 12 test(s) broken

**Comment by arcadia-devtools** [comment]
@annaveronika, internal review request updated: [961303](https://a.yandex-team.ru/review/961303)

**Comment by arcadia-devtools** [comment]
Check [961303](https://a.yandex-team.ru/review/961303) failed: 3 test(s) broken

**Comment by annaveronika** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@annaveronika, internal review request created: [969797](https://a.yandex-team.ru/review/969797)

**Pull request merged**